File: python/cancel_order.py
"""Place a resting order (GTC, 3% below mid), then cancel it."""

import json
from client import exchange, sign_hash, get_mid

COIN = "BTC"

mid = get_mid(COIN)
if mid == 0:
    print(f"Could not fetch {COIN} mid price")
    raise SystemExit(1)

sz = round(11.0 / mid, 5)
rest_px = int(mid * 0.97)

print(f"{COIN} mid: ${mid:,.2f}")
print(f"Placing resting BUY {sz} @ {rest_px} (GTC, 3% below mid)\n")

res = exchange({
    "action": {
        "type": "order",
        "orders": [{
            "asset": COIN,
            "side": "buy",
            "price": str(rest_px),
            "size": str(sz),
            "tif": "gtc",
        }],
    },
})
sig = sign_hash(res["hash"])

result = exchange({
    "action": res["action"],
    "nonce": res["nonce"],
    "signature": sig,
})

exchange_resp = result["exchangeResponse"]
statuses = exchange_resp.get("response", {}).get("data", {}).get("statuses", [])

oid = None
for s in statuses:
    if isinstance(s, dict) and "resting" in s:
        oid = s["resting"].get("oid")
        break

if oid is None:
    print("Could not extract OID from resting order")
    print(json.dumps(exchange_resp, indent=2))
    raise SystemExit(1)

print(f"Order resting (OID: {oid})")
print("Cancelling...\n")

cancel_action = {
    "type": "cancel",
    "cancels": [{"a": COIN, "o": oid}],
}

res = exchange({"action": cancel_action})
sig = sign_hash(res["hash"])

result = exchange({
    "action": cancel_action,
    "nonce": res["nonce"],
    "signature": sig,
})

print(json.dumps(result["exchangeResponse"], indent=2))
print("\nOrder cancelled.")

File: typescript/src/cancelOrder.ts
import { rpc, signHash, getMid } from "./client";

const COIN = "BTC";

async function main() {
  const mid = await getMid(COIN);
  if (mid === 0) {
    console.error(`Could not fetch ${COIN} mid price`);
    process.exit(1);
  }

  const sz = (11.0 / mid).toFixed(5);
  const restPx = Math.floor(mid * 0.97).toString();

  console.log(`${COIN} mid: $${mid.toLocaleString()}`);
  console.log(`Placing resting BUY ${sz} @ ${restPx} (GTC, 3% below mid)\n`);

  const orderAction = {
    type: "order",
    orders: [{ a: COIN, b: true, p: restPx, s: sz, r: false, t: { limit: { tif: "Gtc" } } }],
    grouping: "na",
  };

  let res = await rpc("hl_buildOrder", { action: orderAction });
  let sig = await signHash(res.result.hash);

  const result = await rpc("hl_sendOrder", {
    action: res.result.action || orderAction,
    nonce: res.result.nonce,
    signature: sig,
  });

  const exchange = result.result.exchangeResponse;
  const statuses = exchange?.response?.data?.statuses || [];

  let oid: number | null = null;
  for (const s of statuses) {
    if (s.resting) {
      oid = s.resting.oid;
      break;
    }
  }

  if (oid === null) {
    console.error("Could not extract OID from resting order");
    console.log(JSON.stringify(exchange, null, 2));
    process.exit(1);
  }

  console.log(`Order resting (OID: ${oid})`);
  console.log("Cancelling...\n");

  const cancelAction = {
    type: "cancel",
    cancels: [{ a: COIN, o: oid }],
  };

  res = await rpc("hl_buildCancel", { action: cancelAction });
  sig = await signHash(res.result.hash);

  const cancelResult = await rpc("hl_sendCancel", {
    action: cancelAction,
    nonce: res.result.nonce,
    signature: sig,
  });

  console.log(JSON.stringify(cancelResult.result.exchangeResponse, null, 2));
  console.log("\nOrder cancelled.");
}

main();
